pkg/volume/volume.go

```diff
@@ -223,95 +222,133 @@ func (err deletedVolumeInUseError) Error() string {
 	return string(err)
 }
 
-func RenameDirectory(oldPath, newName string) (string, error) {
-	newPath, err := ioutil.TempDir(filepath.Dir(oldPath), newName)
+// RenameWithFallback attempts to rename a file or directory, but falls back to
+// copying in the event of a cross-link device error. If the fallback copy
+// succeeds, src is still removed, emulating normal rename behavior.
+func RenameWithFallback(src, dest string) error {
+	fi, err := os.Lstat(src)
 	if err != nil {
-		return "", err
+		return err
 	}
 
-	// os.Rename call fails on windows (https://github.com/golang/go/issues/14527)
-	// Replacing with copyFolder to the newPath and deleting the oldPath directory
-	if runtime.GOOS == "windows" {
-		err = copyFolder(oldPath, newPath)
-		if err != nil {
-			glog.Errorf("Error copying folder from: %s to: %s with error: %v", oldPath, newPath, err)
-			return "", err
+	// If renaming a directory make sure the destination does not exist.
+	if fi.IsDir() {
+		if _, err := os.Stat(dest); !os.IsNotExist(err) {
+			if err := os.RemoveAll(dest); err != nil {
+				return err
+			}
 		}
-		os.RemoveAll(oldPath)
-		return newPath, nil
 	}
 
-	err = os.Rename(oldPath, newPath)
-	if err != nil {
-		return "", err
+	err = os.Rename(src, dest)
+	if err == nil {
+		return nil
 	}
-	return newPath, nil
+
+	terr, ok := err.(*os.LinkError)
+	if !ok {
+		return err
+	}
+
+	// Rename may fail if src and dest are on different devices; fall back to
```

Comment:
is this hypothetical or did we experience it?  Since we're renaming it within the same parent dir, this seems impossible to hit?  i.e. `/foo/bar/bat -> /foo/bar/qux` - how could that be XDEV?